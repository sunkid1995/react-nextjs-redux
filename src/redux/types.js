export const PREFIXES = {
  AUTH: 'auth/',
  SERVICE_API: 'serviceApi/',
  SHARED: 'shared/',
};

export const INIT_ACTION = { type: '@@redux/INIT' };

export const AUTH = {
  AUTHORIZE: 'auth/AUTHORIZE',
  DEAUTHORIZE: 'auth/DEAUTHORIZE',
};

export const SERVICE_API = {
  AUTH_WITH_EMAIL: 'serviceApi/AUTH_WITH_EMAIL',
  FORGOT_PASSWORD: 'serviceApi/FORGOT_PASSWORD',
  RESET_PASSWORD: 'serviceApi/RESET_PASSWORD',
};

export const SHARED = {

};
