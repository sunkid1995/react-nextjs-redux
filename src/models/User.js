import Joi from 'joi-browser';
import _ from 'lodash';

// Models
import BaseModel, { COMMON_SCHEMAS } from './BaseModel';

const { ID, STRING, STRINGS_ARRAY } = COMMON_SCHEMAS;

const ROLES = {
  MANAGER: 'sales_manager',
  STAFF: 'sales_man',
  USER: 'User',
};

const { MANAGER, STAFF } = ROLES;

export default class User extends BaseModel {
  static get klass() { return 'User'; }
  
  static dataValidator(): Joi.SchemaMap {
    return {
      auth_token: STRING.required(),
      display_name: STRING.required(),
      email: STRING.required(),
      id: ID,
      nickname: STRING.required(),
      picture: STRING,
      roles: STRINGS_ARRAY,
    };
  }

  get isManager() { return _.includes(this.roles, MANAGER); }
  get isStaff() { return _.includes(this.roles, STAFF) && !_.includes(this.roles, MANAGER); }
}
