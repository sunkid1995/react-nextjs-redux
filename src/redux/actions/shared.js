// import { createAction } from 'redux-actions';

// // Redux
// import { SHARED } from 'src/redux/types';
