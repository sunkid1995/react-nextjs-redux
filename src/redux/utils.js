import _ from 'lodash';

import { IN_DEV_MODE } from 'src/constants';

export function buildErrorMessage(errors) {
  if (_.isEmpty(errors)) return null;

  const transformer = ({ code, message }) => (
    IN_DEV_MODE && code != null ? `${code} - ${message}` : message
  );
  return _.compact(_.map(errors, transformer)).join(', ');
}

export function buildHeaders(state) {
  const { auth: { user: { auth_token: token } = {} } } = state;
  const bearerToken = token != null ? `Bearer ${token}` : undefined;

  return {
    Accept: 'application/json',
    Authorization: bearerToken,
  };
}

export function responseInterceptor(response, callback) {
  if (response == null) return;
  return callback(response);
}
