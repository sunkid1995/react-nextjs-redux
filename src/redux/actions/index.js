import * as auth from './auth';
import * as serviceApi from './serviceApi';
import * as shared from './shared';

export { auth, serviceApi, shared };
