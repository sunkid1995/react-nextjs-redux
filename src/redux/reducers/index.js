import auth from './auth';
import meta from './meta';
import serviceApi from './serviceApi';
import shared from './shared';

export { auth, meta, serviceApi, shared };
