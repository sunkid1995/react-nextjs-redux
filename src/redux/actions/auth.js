import { createAction } from 'redux-actions';

// Redux
import { AUTH } from 'src/redux/types';

export const authorize = createAction(AUTH.AUTHORIZE);
export const deauthorize = createAction(AUTH.DEAUTHORIZE);
