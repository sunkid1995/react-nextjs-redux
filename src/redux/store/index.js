import { applyMiddleware, compose, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { persistCombineReducers, persistStore } from 'redux-persist';
import reduxThunk from 'redux-thunk';

import { IN_DEV_MODE } from 'src/constants';
import { auth, meta, serviceApi, shared } from 'src/redux/reducers';

// Locals
import buildAxiosMiddleware from './buildAxiosMiddleware';
import buildPersistorConfigs from './buildPersistorConfigs';

const composeEnhancer = IN_DEV_MODE ? composeWithDevTools({}) : compose;

export default function makeStore(/* initialState, options */) {
  const configs = buildPersistorConfigs();

  const store = createStore(
    persistCombineReducers(configs, { auth, meta, serviceApi, shared }),
    composeEnhancer(applyMiddleware(buildAxiosMiddleware(), reduxThunk)),
  );

  persistStore(store);
  return store;
}
