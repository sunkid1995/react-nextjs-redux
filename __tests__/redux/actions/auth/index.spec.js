import { auth } from 'src/redux/actions';

// Models
import User from 'src/models/User';

describe('redux/actions/auth', () => {
  const user = User.build(require('src/data/fixtures/user.json'));

  test('authorize', () => {
    expect(auth.authorize(user)).toEqual({ type: 'auth/AUTHORIZE', payload: user });
    expect(auth.authorize(user)).toMatchSnapshot();
  });

  test('deauthorize', () => {
    expect(auth.deauthorize()).toEqual({ type: 'auth/DEAUTHORIZE' });
    expect(auth.deauthorize()).toMatchSnapshot();
  });
});
