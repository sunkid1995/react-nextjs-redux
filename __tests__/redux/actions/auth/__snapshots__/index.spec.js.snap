// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/actions/auth authorize 1`] = `
Object {
  "payload": User {
    "auth_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6InRoYW5obnhAb25hY2xvdmVyLmNvbSIsInVpZCI6IjgzOWU2OGJkLTJjN2QtNDkzMy1hMmI3LWYxZDM0MDkyODI0NiIsImlhdCI6MTUyMDg0MjA3MCwiZXhwIjoxNTIxNDQ2ODcwLCJqdGkiOiJiMDlhYjM4NS1iZWZiLTRiNzYtODQyZC0xODQ1ZWU2MGJhNGEifQ.zUxEMCY2zYj6y4jw6-sJAHig-hPzDY_l_MoCT4XfbhI",
    "display_name": "ThanhNX",
    "email": "thanhnx@onaclover.com",
    "id": 4,
    "nickname": "ThanhNX",
    "roles": Array [
      "user",
      "sales_manager",
      "sales_man",
    ],
  },
  "type": "auth/AUTHORIZE",
}
`;

exports[`redux/actions/auth deauthorize 1`] = `
Object {
  "type": "auth/DEAUTHORIZE",
}
`;
